### PR TITLE
Implement Subdocument feature

### PR DESCRIPTION
**Pull request by pasin**
* Any Dictionaries are mapped to CBLSubdocument.
* When document is saved, update all subdocuments with new roots or invalidate them if they do not valid anymore.

**Discussion Points:**
* Current the invalidated subdocuments are also removed from cache (_properties) so they are no longer getting any new updates.

### CODE REVIEW

File: Objective-C/CBLDatabase.mm [hunk 1]
```diff
@@ -290,6 +291,13 @@ - (void) document: (CBLDocument*)doc hasUnsavedChanges: (bool)unsaved {
 }
 
 
+- (NSMapTable*) sharedStrings {
+    if (!_sharedStrings)
+        _sharedStrings = FLCreateSharedStringsTable();
```
Comment (snej) [inline Objective-C/CBLDatabase.mm]: This NSMapTable is only to be used with a single Fleece document (root FLValue.) The docs and naming are misleading; I think I told you the wrong thing earlier because I thought you were talking about the per-database shared string table. Sorry!
Just take this out for now -- it's just an optimization. We can put it in later.

File: Objective-C/CBLProperties.mm [hunk 2]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
 
 - (nullable id) objectForKeyedSubscript: (NSString*)key {
-    if (_properties)
-        return _properties[key];
-    else
-        return [self fleeceValueToObject: [self fleeceValueForKey: key]];
+    id obj = _properties ? _properties[key] : nil;
+    if (!obj) {
+        auto value = [self fleeceValueForKey: key];
+        obj = [self fleeceValueToObject: value  forKey: key asJson: NO];
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: You can check for `value==nullptr` here and return nil, since it's a common case, and the other two calls are just no-ops.

File: Objective-C/CBLProperties.mm [hunk 3]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+
+
+- (void) cacheValue: (nullable id)value forKey: (NSString*)key changed: (BOOL)changed {
+    BOOL shouldCache = changed ||
+                       [value isKindOfClass: [CBLSubdocument class]] ||
+                       [value isKindOfClass: [NSArray class]];
+    if (shouldCache) {
+        if (!_properties)
+            _properties = [NSMutableDictionary dictionary];
+        _properties[key] = value;
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: This will crash if value is nil.
An alternative is `[_properties setValue: value forKey: key]`, which allows a nil value.

File: Objective-C/CBLProperties.mm [hunk 4]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
 - (void) setObject: (nullable id)value forKeyedSubscript: (NSString*)key {
-    [self mutateProperties];
-    
     // NSDate:
     if ([value isKindOfClass: NSDate.class])
         value = [CBLJSON JSONObjectWithDate: value];
     
     id oldValue = _properties[key];
     
     if (![value isEqual: oldValue] && value != oldValue) {
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: Since `_properties` is a cache, this method will do the wrong thing if called with a nil `value`, in a document where the key has a value but it hasn't been accessed yet. `oldValue` will be nil, so the `if` block will be skipped and nothing will change.

I think you need to change `_properties[key]` to `self[key]`.

File: Objective-C/CBLProperties.mm [hunk 5]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+            subdoc = [self subdocumentWithRoot: nullptr withProperties: value forKey: key];
+        return subdoc;
+    } else if ([value isKindOfClass: [NSArray class]]) {
+        NSArray* oldArray = $castIf(NSArray, oldValue);
+        NSArray* nuArray = (NSArray*)value;
+        NSMutableArray* result = [NSMutableArray arrayWithCapacity: nuArray.count];
+        for (int i = 0; i < nuArray.count; i++) {
+            id oldValue = oldArray && i < oldArray.count ? oldArray[i] : nil;
+            id nuValue = nuArray[i];
+            result[i] = [self convertValue: nuValue oldValue: oldValue forKey: key];
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: This assumes the old and new values at the same indexes correspond to each other. But that's not the case if items have been inserted or removed in the middle of the array. Or if the two arrays are entirely different.

File: Objective-C/CBLProperties.mm [hunk 6]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+                [oldSubdoc invalidate];
+            }
+        } else {
+            CBLSubdocument* oldSubdoc = $castIf(CBLSubdocument, oldValue);
+            [oldSubdoc invalidate];
+        }
+        return result;
+    } else {
+        CBLSubdocument* nuSubdoc = $castIf(CBLSubdocument, value);
+        [self attachSubdocument: nuSubdoc forKey: key];
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: This will blow up if `value` isn't a `CBLSubDocument`, i.e. it's an NSString, NSNumber, NSNull. -attachSubdocument doesn't allow a nil first parameter.

File: Objective-C/CBLProperties.mm [hunk 7]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+        
+        if ([oldValue isKindOfClass: [CBLSubdocument class]])
+            [oldValue invalidate];
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: What about if nuSubDoc == oldSubDoc?

File: Objective-C/CBLProperties.mm [hunk 8]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+            FLDictIterator_Begin(_root, &iter);
+            do {
+                NSString* key = [self fleeceValueToKeyString: FLDictIterator_GetKey(&iter)];
+                id value = _properties[key];
+                if (value) {
+                    if (asJson)
+                        value = [self jsonValueForValue: value];
+                    [changedNames removeObject: key];
+                } else {
+                    value = [self fleeceValueToObject: FLDictIterator_GetValue(&iter)
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: This doesn't work for the case of a property that's been deleted (but not yet saved). In this case the property will have a value in `_root`, but not in `_properties`. (And the key will be in `_changedNames`.)
I think the right algorithm to use is
1. Set the result to a mutable copy of `_properties`
2. Iterate through `_root`.
   1. If the key is not in `_changedKeys`, and there's no value for the key in `_properties`, decode the Fleece value and add it to the result.

File: Objective-C/CBLProperties.mm [hunk 9]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+            id value = _properties[key];
+            result[key] = asJson ? [self jsonValueForValue: value] : value;
+        }
+    }
+    return result;
+}
+
+
+- (id) jsonValueForValue: (id)value {
+    if ([value conformsToProtocol: @protocol(CBLJSONEncoding)]) {
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: NSDate?

File: Objective-C/CBLProperties.mm [hunk 10]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
 
-- (void) mutateProperties {
-    if (!_properties)
-        _properties = [self.savedProperties mutableCopy];
+- (void) markChangedKeys: (NSArray*)keys {
+    if (!_changedNames)
+        _changedNames = [NSMutableSet set];
+    
+    for (NSString* key in keys) {
+        if (_properties[key] != nil && ![_changedNames containsObject: key])
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: You don't need to call -containsObject:. -addObject: will do the right thing.

File: Objective-C/CBLProperties.mm [hunk 11]
```diff
@@ -22,85 +22,84 @@
...
 
 
 - (BOOL) booleanForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return [numberProperty(_properties ? _properties : _rootProps, key) boolValue];
-    else
-        return FLValue_AsBool([self fleeceValueForKey: key]);
-    return NO;
+    id v = _properties ? _properties[key] : nil;
+    return v ? [numberProperty(v) boolValue] : FLValue_AsBool([self fleeceValueForKey: key]);
```
Comment (pasin) [inline Objective-C/CBLProperties.mm]: ok, I will make it consistent with the lite-core or fleece.

File: Objective-C/CBLDocument.mm [hunk 12]
```diff
@@ -106,7 +106,7 @@ - (BOOL) purge: (NSError**)outError {
 
 
 - (void) revert {
-    [self resetChanges];
+    [self revertChanges];
```
Comment (pasin) [inline Objective-C/CBLDocument.mm]: I will make the change.

File: Objective-C/CBLProperties.mm [hunk 13]
```diff
@@ -22,85 +22,84 @@
...
-        return slice2string((FLValue_AsString([self fleeceValueForKey: key])));
+    id v = _properties ? _properties[key] : nil;
+    return v ? $castIf(NSString, v) : slice2string((FLValue_AsString([self fleeceValueForKey: key])));
+}
+
+
+- (CBLSubdocument*) subdocumentForKey: (NSString*)key {
+    id subdoc = self[key];
+    if (!subdoc) {
+        subdoc = [self subdocumentWithRoot: nullptr withProperties: nil forKey: key];
```
Comment (pasin) [inline Objective-C/CBLProperties.mm]: By the API design, -subdocumentForKey: has the same behavior as Database's -documentWithID:, which will do get-or-create. We could discuss about next week. I will setup a meeting.

File: Objective-C/CBLProperties.mm (hunk 5 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: I don't have a good solution to match subdocuments in two arrays. From above, it just tries to match as possible. This is another thing that I want to discuss next week.

File: Objective-C/CBLProperties.mm (hunk 6 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: Yep.

File: Objective-C/CBLProperties.mm (hunk 7 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: Yep, I missed that case.

File: Objective-C/CBLProperties.mm (hunk 9 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: In most case, when setting a NSDate object, it will get converted to string. BUT there is one case that the NSDate objects will not get converted are when they are in an array.

File: Objective-C/CBLProperties.mm (hunk 3 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: When I tested it, it didn't crash. It just removed the value of the key.

File: Objective-C/CBLProperties.h [hunk 14]
```diff
@@ -44,6 +44,9 @@ NS_ASSUME_NONNULL_BEGIN
 /** Get an NSString property by key. Returns nil if the property doesn't exists. */
 - (nullable NSString*) stringForKey: (NSString*)key;
 
+/** Get or create a new CBLSubdocument by key. */
```
Comment (pasin) [inline Objective-C/CBLProperties.h]: Will discuss about this next week.

File: Objective-C/CBLProperties.mm [hunk 15]
```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
+                return [self subdocumentWithRoot: (FLDict)value withProperties: nil forKey:key];
+            }
+            case kFLArray: {
+                FLArray array = FLValue_AsArray(value);
+                int count = FLArray_Count(array);
+                NSMutableArray* result = [[NSMutableArray alloc] initWithCapacity: count];
+                if (count > 0) {
+                    FLArrayIterator iter;
+                    FLArrayIterator_Begin(array, &iter);
+                    do {
```
Comment (pasin) [inline Objective-C/CBLProperties.mm]: OK

File: Objective-C/CBLProperties.mm (hunk 8 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: I thought the deleted case got taken care after iterating loop. I will revise the logic. Thanks.

File: Objective-C/CBLProperties.mm (hunk 15 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: Actually there is an if check if count > 0 then doing the loop, so I think I will keep the code the same as it looks cleaner than checking if FLArrayIterator_GetValue return nullprt.

**Comment by pasin** [comment]
I fixed the code except the ones that are related to the API that we will discuss next week. I also rebased with the current feature/2.0 branch to get all of the new configuration.

File: Objective-C/CBLProperties.mm (hunk 3 above)
Comment (snej) [inline Objective-C/CBLProperties.mm]: Hey, you're right — the API docs for [NSMutableDictionary](https://developer.apple.com/reference/foundation/nsmutabledictionary/1574187-setobject) say that the subscript operator allows a nil value, unlike -setObject:forKey:. I didn't know that!

**Comment by pasin** [comment]
@snej I have made changes according to our conversation today (Commit 9e56c3dd004c5b9416ca6e81c0ad020fa210eeed). I also rebased with the current feature/2.0. Could you please re-review? Thanks.

File: Objective-C/CBLProperties.mm [hunk 16]
```diff
@@ -22,85 +22,83 @@
...
 
 
 - (BOOL) booleanForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return [numberProperty(_properties ? _properties : _rootProps, key) boolValue];
-    else
+    id v = _properties[key];
+    if (v) {
+        id n = numberProperty(v);
+        return n ? [n boolValue] : YES;
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: NSNull should also return NO

File: Objective-C/CBLProperties.mm [hunk 17]
```diff
@@ -22,85 +22,83 @@
...
+    } else
         return FLValue_AsBool([self fleeceValueForKey: key]);
-    return NO;
 }
 
 
 - (nullable NSDate*) dateForKey: (NSString*)key {
     NSString* dateStr = [self stringForKey: key];
-    return [CBLJSON dateWithJSONObject: dateStr];
+    return dateStr ? [CBLJSON dateWithJSONObject: dateStr] : nil;
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: You don't need to test for nil; +dateWithJSONObject already handles that.

File: Objective-C/CBLProperties.mm [hunk 18]
```diff
@@ -22,85 +22,83 @@
...
 }
 
 
 - (nullable NSString*) stringForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return _properties ? _properties[key] : _rootProps[key];
-    else
-        return slice2string((FLValue_AsString([self fleeceValueForKey: key])));
+    id v = self[key];
+    return v ? $castIf(NSString, v) : nil;
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: Don't need to test for nil; `$castIf` already handles that.

File: Objective-C/CBLProperties.mm [hunk 19]
```diff
@@ -22,85 +22,83 @@
...
+}
+
+
+- (nullable CBLSubdocument*) subdocumentForKey: (NSString*)key {
+    id v = self[key];
+    return v ? $castIf(CBLSubdocument, v) : nil;
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: Don't need to test for nil; `$castIf` already handles that.

File: Objective-C/CBLProperties.mm [hunk 20]
```diff
@@ -146,101 +144,346 @@ - (BOOL) containsObjectForKey: (NSString*)key {
 
 
 - (nullable id) objectForKeyedSubscript: (NSString*)key {
-    if (_properties)
-        return _properties[key];
-    else
-        return [self fleeceValueToObject: [self fleeceValueForKey: key]];
+    id obj = _properties ? _properties[key] : nil;
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: Don't need to check for nil.

File: Objective-C/CBLProperties.mm [hunk 21]
```diff
@@ -146,101 +144,346 @@ - (BOOL) containsObjectForKey: (NSString*)key {
...
     }
 }
 
 
-#pragma mark - INTERNAL:
+- (void) revert {
+    for (NSString* key in _changedNames) {
+        id value = _properties[key];
+        if ([value isKindOfClass: [CBLSubdocument class]]) {
+            CBLSubdocument* subdoc = $cast(CBLSubdocument, value);
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: `$cast` will do another isKindOf: test, which is redundant; just do a C cast instead. (Or do a `$castIf` and assign it to subdoc, then test if subdoc is nil; that's the pattern I usually use.)

File: Objective-C/CBLProperties.mm [hunk 22]
```diff
@@ -22,85 +22,83 @@
...
+    return [self currentPropertiesAsJson: NO];
 }
 
 
 - (void) setProperties: (NSDictionary*)properties {
-    _properties = properties ? [properties mutableCopy] : [NSMutableDictionary dictionary];
-    [self markChanges];
+    _properties = [self convertProperties: properties];
+    _changedNames = nil;
+    [self markChangedKeys: [_properties allKeys]];
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: I don't think this will do the right thing for keys that are in `_root` but are not in the new `properties`. They should be added to `_changedNames`, but they won't be.

File: Objective-C/CBLProperties.mm [hunk 23]
```diff
@@ -146,101 +144,346 @@ - (BOOL) containsObjectForKey: (NSString*)key {
 
 
 - (nullable id) objectForKeyedSubscript: (NSString*)key {
-    if (_properties)
-        return _properties[key];
-    else
-        return [self fleeceValueToObject: [self fleeceValueForKey: key]];
+    id obj = _properties ? _properties[key] : nil;
+    if (!obj) {
```
Comment (snej) [inline Objective-C/CBLProperties.mm]: This method doesn't handle removed properties. If a property exists in `_root`, and then I set its value to `nil`, then this method _should_ return the new value of `nil`. But this code will see that there's no value in `_properties`, and fall back to reading the value from Fleece.
This line should also be checking whether `key` is in `_changedNames`, and if it is, skipping the following block. I.e. `if (!obj && ![_changedNames containsObject: key])`.
I don't know if you've looked at the property-handling code of CBLModel, but it does this.

File: Objective-C/CBLProperties.mm (hunk 22 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: I totally missed this. Thanks!

File: Objective-C/CBLProperties.mm (hunk 23 above)
Comment (pasin) [inline Objective-C/CBLProperties.mm]: Again, I totally missed the remove case. Thanks!

**Comment by pasin** [comment]
@snej I made changes per your comments and push. Thanks a lot for the review.

**Comment by pasin** [comment]
I just fixed the conflicts and a broken integration with notification after the merge.

**Comment by pasin** [comment]
As the feature/2.0 branch has been changed a lot, I'm taking the PR back and will resubmit after reimplementing.
